### PR TITLE
add CI jobs to implement platform support RFC

### PR DESCRIPTION
**Pull request by brianlheim**
## Purpose and Motivation

part of implementing https://github.com/supercollider/rfcs/pull/1 was adding CI jobs to enforce compatibility with platforms and toolchains, this PR does that.

no new Windows jobs because each new job adds ~15 minutes to the total build time.

it's possible that Travis builds will take too long after we start doing this; we can remove jobs if that's the case. we can also run some of them less often, for instance on cron schedules or when a special keyword is in the commit message or when making a release.

## Summary of changes

- `.travis.yml`: rewrite build matrix and add new jobs for a total of 18. this covers a variety of toolchains and build configurations; see the "summary" comment at the top of that file for full details
- `.travis.yml`: add some logic around deployments to ensure only 1 macOS build is deployed
- `before-install-linux.sh`, `before-install-osx.sh`: update logic to install the right packages depending on configuration
- `before-script-linux.sh`, `before-script-osx.sh`: use correct CMake flags depending on build job configuration
- `script-linux.sh`: use 2 cores when building on linux, which we already do in sc3-plugins with no problems encountered so far
- `CMakeLists.txt`: add an option for using libstdc++ with clang, `SC_CLANG_USES_LIBSTDCPP`; remove obsolete logic for pre-3.1 cmake.
- `README_LINUX.md`: add info on using clang
- `README_MACOS.md`: update the version of Xcode required to 10
- `server/supernova/CMakeLists.txt`: link against libatomic on clang version 4, 5, 6; a build issue i encountered when running clang build jobs
- other C++ files: expand a workaround we previously added for appleclang 11 to include appleclang 10 and 11, and clang 7 and 8.
- rename the VERSION at `external_libraries/libsndfile/VERSION` to VERSION.txt; very obscure issue. with Boost 1.73 on macOS (which happens in the system libs build), a boost header ends up including this file because it thinks it's the C++20 [`<version>` header](https://en.cppreference.com/w/cpp/utility#Language_support). i don't like modifying 3rd party code but it doesn't seem like this file even exists in the main repo, and it's not used by our build system. renaming it seemed to be the most sensible solution to a really nonsensical problem.

## Types of changes

- Documentation - updated readmes
- Bug fix - bug fixes for clang and xcode versions
- New feature - new ci jobs and cmake options

## To-do list

- [x] Code is tested
- [ ] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by brianlheim** [comment]
some, if not all the linux failures are because i turned on scel, but the PR to update the scel submodule hasn't been merged into develop yet.

**Comment by brianlheim** [comment]
i've added 3 new jobs:
- gcc10 on ubuntu (passing)
- LIBSCSYNTH=ON on ubuntu (failing, but should pass, investigating)
- LIBSCSYNTH=ON on ubuntu (failing, and should fail until #5013 is fixed, investigating)

File: common/SC_AppleEventLoop.mm [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    // because we don't want to always show an icon in the docker.
+    // ProcessSerialNumber psn = { 0, kCurrentProcess };
+    // TransformProcessType(&psn, kProcessTransformToForegroundApplication);
+
+    // Create NSApplication
+    [NSApplication sharedApplication];
+}
+
+void run() {
+#if 0
```
Comment (joshpar) [inline common/SC_AppleEventLoop.mm:41]: Since it looks like this has been if'd out forever now, do you feel ok if we just get rid of it?

Comment (brianlheim) [reply common/SC_AppleEventLoop.mm:41]: yeah! but that code is actually just to demonstrate PR #5014 -- i'll make that change there :)

Comment (brianlheim) [reply common/SC_AppleEventLoop.mm:41]: (done - a8657956b0)

**Comment by brianlheim** [comment]
@joshpar i updated this PR after #5014. anything else you'd like to add?

**Comment by brianlheim** [comment]
oh, i just realized i haven't touched the deployment aspect of this yet. we don't want all those macOS jobs trying to push to S3 or GitHub.

**Comment by brianlheim** [comment]
ready for review again. i just pushed this to the branch on this repo as well to see if the deployment step will work correctly.

**Comment by brianlheim** [comment]
i think i've finally resolved all the issues here. the PR text is updated with a summary of the changes. @joshpar (or anyone else) can you please review when you have time?

**Review by dyfer — APPROVED** [review]
looks great, thanks!

**Comment by brianlheim** [comment]
thanks @dyfer !

**Pull request merged**